diesel/src/sqlite/types/json.rs

```diff
@@ -0,0 +1,575 @@
+        let (value, value_consumed) = read_jsonb_value(&bytes[total_read..payload_size])?;
+        total_read += value_consumed;
+
+        object.insert(key_str, value);
+    }
+
+    if total_read != payload_size {
+        return Err("Object payload size mismatch".into());
+    }
+
+    Ok(serde_json::Value::Object(object))
+}
+
+fn write_jsonb_header(
+    buffer: &mut Vec<u8>,
+    element_type: u8,
+    payload_size: usize,
+) -> serialize::Result {
+    let payload_size = u8::try_from(payload_size).map_err(Box::new)?;
+
+    buffer.push((payload_size << 4) | element_type);
+
+    Ok(IsNull::No)
+}
+
+// Helper function to write a JSON value into a JSONB binary format
+fn write_jsonb_value(value: &serde_json::Value, buffer: &mut Vec<u8>) -> serialize::Result {
+    match value {
+        serde_json::Value::Null => write_jsonb_null(buffer),
+        serde_json::Value::Bool(b) => write_jsonb_bool(*b, buffer),
+        serde_json::Value::Number(n) => write_jsonb_number(n, buffer),
+        serde_json::Value::String(s) => write_jsonb_string(s, buffer),
+        serde_json::Value::Array(arr) => write_jsonb_array(arr, buffer),
+        serde_json::Value::Object(obj) => write_jsonb_object(obj, buffer),
+    }
+}
+
+// Write a JSON null
+fn write_jsonb_null(buffer: &mut Vec<u8>) -> serialize::Result {
+    write_jsonb_header(buffer, JSONB_NULL, 0x0)?;
+    Ok(IsNull::No)
+}
+
+// Write a JSON boolean
+fn write_jsonb_bool(b: bool, buffer: &mut Vec<u8>) -> serialize::Result {
+    // Use the constants for true and false
+    write_jsonb_header(buffer, if b { JSONB_TRUE } else { JSONB_FALSE }, 0x0)?;
+    Ok(IsNull::No)
+}
+
+// Write a JSON number (integers and floats)
+fn write_jsonb_number(n: &serde_json::Number, buffer: &mut Vec<u8>) -> serialize::Result {
+    if let Some(i) = n.as_i64() {
+        // Write an integer (INT type)
+        write_jsonb_int(i, buffer)
+    } else if let Some(f) = n.as_f64() {
+        // Write a float (FLOAT type)
+        write_jsonb_float(f, buffer)
+    } else {
+        Err("Invalid number type".into())
+    }
+}
+
+// Write an integer in JSONB format
+fn write_jsonb_int(i: i64, buffer: &mut Vec<u8>) -> serialize::Result {
+    let int_str = i.to_string();
+
+    write_jsonb_header(buffer, JSONB_INT, int_str.len())?;
+
+    // Write the ASCII text representation of the integer as the payload
+    buffer.extend_from_slice(int_str.as_bytes());
+
+    Ok(IsNull::No)
+}
+
+// Write a floating-point number in JSONB format
+fn write_jsonb_float(f: f64, buffer: &mut Vec<u8>) -> serialize::Result {
+    let float_str = f.to_string();
+
+    write_jsonb_header(buffer, JSONB_FLOAT, float_str.len())?;
+
+    // Write the ASCII text representation of the float as the payload
+    buffer.extend_from_slice(float_str.as_bytes());
+
+    Ok(IsNull::No)
+}
+
+// Write a JSON string
+fn write_jsonb_string(s: &str, buffer: &mut Vec<u8>) -> serialize::Result {
+    write_jsonb_header(buffer, JSONB_TEXT, s.len())?;
```

Comment:
It might be meaningful to use `JSONB_TEXTRAW` here as we don't handle any escaping. That seems like the easiest solution, although I'm not sure if that has any downsides in compatibility to other tools. That might require some additional tests to confirm that at least sqlite can handle that.

Reply:
> It might be meaningful to use JSONB_TEXTRAW here as we don't handle any escaping.

@weiznich do you have an actual example of a valid Rust String that would be interpreted as TEXTRAW by SQLite?

Reply:
The main problem is that I don't think `serde_json::Value` deals only with `String`. So my understanding is that there is no `String` value that would be written as a `TEXTRAW` field in SQLite's JSONB. But I might be wrong.

If I am right, the `TEXTRAW` is out of scope for this MR that revolves solely around `serde_json::Value`.

Reply:
My idea was to not bother with deciding if we can write something as `JSONB_TEXT` or not and always mark the buffer as `TEXTRAW`. As you already implemented the better more complete variant that's also fine and even preferable.